### ISSUE TITLE
reproduce the results

### ISSUE THREAD

**Issue by sc07231** [antonio-mastropaolo/LANCE#4]
Hello, dear authors:

In the code you released, I haven’t seen the code for predicting results. Here is my way to reproduce the results with the fine-tuned model provided by you. First, I will transfer the TensorFlow fine-tuned model to the PyTorch version, and then use the following code with huggingface’s transformers library to predict results:

```
batch_size = 1
beam_num = 1
input_max_length = 512
output_max_length = 512
device = 'cuda'
model_dir = 'lance-model/log-inject-pt’
sp_path = 'SP_LOG.model'
input_file = ’test.csv'

input_texts = pd.read_csv(input_file)
input_texts = input_texts['Input']
total_len = len(input_texts)

tokenizer = T5Tokenizer.from_pretrained(sp_path)
model = T5ForConditionalGeneration.from_pretrained(model_dir)
model.to(device)
model.eval()

torch.manual_seed(0)
predictions = []

with torch.no_grad():
    cache = []

    for index, input_text in input_texts.items():
        if index > 0 and index % 100 == 0:
            print("Progress: {}/{}".format(index, total_len), len(predictions), predictions[-1])

        cache.append(input_text)
        if (index + 1) % batch_size != 0 and index + 1 != len(input_texts):
            continue

        tokenized_text = tokenizer(cache, truncation=True, padding='max_length', max_length=input_max_length,
                                   return_tensors='pt')

        source_ids = tokenized_text['input_ids'].to(device, dtype=torch.long)
        source_mask = tokenized_text['attention_mask'].to(device, dtype=torch.long)

        generated_ids = model.generate(
            input_ids=source_ids,
            attention_mask=source_mask,
            max_length=output_max_length,
            num_beams=beam_num,
            num_return_sequences=1,
            early_stopping=True
        )

        for i, generated_id in enumerate(generated_ids):
            generated_text = tokenizer.decode(generated_id, skip_special_tokens=True)
            predictions.append(generated_text)

        cache = []
```

Is that right? I am not sure whether I am doing it the right way cause I can not reproduce the results reported in your paper.
Thank you very much!

**Comment by antonio-mastropaolo** [comment]
@sc07231 

If you want to reproduce our work to its totality, the code to generate predictions is provided within the fine-tuning script.
I'm posting it below for simplicity.

```
# %%capture
model.bach_size=32
model.eval(
    mixture_or_task_name="task",
    checkpoint_steps=-1 
)
```

As a reference, you can inspect the very last cell of the following file (as mentioned above):  https://github.com/antonio-mastropaolo/LANCE/blob/main/Code/Fine-tuning/Fine_Tuning.ipynb

Unfortunately, I can't tell if you achieve similar performances to ours with your implementation. 
Previously, when working on different tasks, I observed a slight decrease in performance when transitioning from TF to HF and subsequently conducting inference on the latter.

Anyhow, the code looks good to me.

**Comment by zhipeng-cai** [comment]
hello, I have used the prediction code in [https://github.com/antonio-mastropaolo/LANCE/blob/main/Code/Fine-tuning/Fine_Tuning.ipynb](https://github.com/antonio-mastropaolo/LANCE/blob/main/Code/Fine-tuning/Fine_Tuning.ipynb) to generate the results, but there are some problems confusing me.

Here is the specific prediction code in the final cell:
```
model.batch_size = 256
input_file = '/content/drive/MyDrive/log//test.tsv'#@param { type: "string" }
output_file = '/content/drive/MyDrive/log/prediction.tsv'#@param { type: "string" }
model.predict(input_file, output_file, checkpoint_steps=500000, vocabulary=get_default_vocabulary())
```

Here is one example of the generated text:
`b'public static Optional < Geometry > wktToGeometry ( String wkt , WKTReader wktReader ) { try { return Optional . of ( wktReader . read ( wkt ) ) ; } catch ( ParseException e ) { LOGGER . debug ( ""unable to convert WKT to a Geometry object: wkt={}" , wkt , e ) ; } return Optional . empty ( ) ; } "public static Optional < Geometry > wktToGeometry ( String wkt , WKTReader wktReader ) { try { return Optional . of ( wktReader . read ( wkt ) ) ; } catch ( ParseException e ) { LOGGER . debug ( ""unable to convert WKT to a Geometry object: wkt={}" , wkt , e ) ; } return Optional . empty ( ) ; }"'`

The generated text is consisted of the input text and output text and the two texts sometimes are quoted by "" and sometimes not, I wonder if this is the right format for the prediction outputs.

And I found some generated output text is not complete, which has been truncated by the right side, how to generate the complete output text?

Thank you very much!

**Comment by zhipeng-cai** [comment]
hello, I have used the prediction code in https://github.com/antonio-mastropaolo/LANCE/blob/main/Code/Fine-tuning/Fine_Tuning.ipynb to generate the results, but there are some problems confusing me.

Here is the specific prediction code in the final cell:

```
model.batch_size = 256
input_file = '/content/drive/MyDrive/log//test.tsv'#@param { type: "string" }
output_file = '/content/drive/MyDrive/log/prediction.tsv'#@param { type: "string" }
model.predict(input_file, output_file, checkpoint_steps=500000, vocabulary=get_default_vocabulary())
```
Here is one example of the generated text:

> b'public static Optional < Geometry > wktToGeometry ( String wkt , WKTReader wktReader ) { try { return Optional . of ( wktReader . read ( wkt ) ) ; } catch ( ParseException e ) { LOGGER . debug ( ""unable to convert WKT to a Geometry object: wkt={}" , wkt , e ) ; } return Optional . empty ( ) ; } "public static Optional < Geometry > wktToGeometry ( String wkt , WKTReader wktReader ) { try { return Optional . of ( wktReader . read ( wkt ) ) ; } catch ( ParseException e ) { LOGGER . debug ( ""unable to convert WKT to a Geometry object: wkt={}" , wkt , e ) ; } return Optional . empty ( ) ; }"'

The generated text is consisted of the input text and output text and the two texts sometimes are quoted by "" and sometimes not, I wonder if this is the right format for the prediction outputs.

And I found some generated output text is not complete, which has been truncated by the right side, how to generate the complete output text?

Thank you very much!

**Comment by antonio-mastropaolo** [comment]
@zhipeng-cai
It appears that you are not utilizing the inference function correctly.
To begin with, it is advisable to employ mode.eval as suggested in the previous response (please refer to the above messages).
Furthermore, if you intend to utilize the predict function, you must provide the model with a text file. This file should contain each Java method that requires the injection of the log statement, with each method listed on a separate line.

To clarify, suppose your test set consists of 5,000 Java Methods. In that case, you would construct a text file similar to the following example:

```
Method_1
Method_2
Method_3
```

As for the quotes, it depends on the input. If the model predicts the method surrounded by quotes, you can easily strip those.

**Comment by JasonBourne1998** [comment]
Hi @zhipeng-cai, have you finished the prediction work? If yes, could you please share the input file? Thanks in advance.
